### PR TITLE
DeepJet b-tagger for run2

### PR DESCRIPTION
**Pull request by panwarlsweet**
1. Integration of DeepJet b-tagger for Run2 analysis
2. Rerun only for 2017 and 2016 samples (only processed with CMSSW_9_4_X and X<10), not on 2018 samples
3. Readme has been updated accordingly.

### CODE REVIEW

**Comment by simonepigazzini** [comment]
Dear Lata,

This branch is supposed to work in CMSSW_10_5_0, did you tested it in this release? Also please remove any "year" switch from the config files and take advantage of the metaConditions json (you can take a look for example of what we do for MET).

Thank you,

simone

**Comment by panwarlsweet** [comment]
Dear Simone,

Yes, I have tested my code with CMSSW_10_5_0. It worked fine. 
And I will update metaConditions as you have suggested to rerun DeepJet. 

Thanks for your feedback.

Lata

File: setup_flashgg.sh [hunk 1]
```diff
@@ -47,6 +47,8 @@ git cms-merge-topic -u simonepigazzini:topic_flashgg_10_5_0_qgtagger
 
 # EGamma post reco tools for energy scales and smearings
 echo "Settinga up Scales and Smearings form EGM"
+git cms-addpkg RecoBTag/TensorFlow
```
Comment (simonepigazzini) [inline setup_flashgg.sh:50]: Can you please move this on other lines (not in between EGM stuff) with an explanation of why we need to cherrypick this particular commit (to help future bookkeeping). Thanks

**Comment by panwarlsweet** [comment]
Dear Simone,

I have added the rerunning DeepJet option for 2016 and 2017 with Metaconditions JSON file and tested my code. It worked fine. Could you please provide your feedback on these changes?

Thanks,

Lata Panwar

File: README.md [hunk 2]
```diff
@@ -39,7 +39,10 @@ Examples to run on RunII legacy test campaign:
 And a very basic workflow test (for reference, this is not supposed to give paper-grade results):
  ```
  cd $CMSSW_BASE/src/flashgg
- cmsRun MicroAOD/test/microAODstd.py processType=sig datasetName=glugluh # or processType=data depending on input file
+cmsRun MicroAOD/test/microAODstd.py processType=sig year=2017 datasetName=glugluh conditionsJSON=MetaData/data/MetaConditions/Era2016_RR-17Jul2018_v1.json
```
Comment (simonepigazzini) [inline README.md]: remove year=2017 from the example to match the changes in the code

**Comment by simonepigazzini** [comment]
Hi Lata,

thank you. It looks good. I made few inline comments on minor points. Can you also cleanup the conflicts by pulling from the official repo?

simone

File: MicroAOD/test/microAODstd.py [hunk 3]
```diff
@@ -82,7 +82,7 @@
 #process.weightsCount.pileupInfo = "addPileupInfo"
 
 from flashgg.MicroAOD.flashggMicroAODOutputCommands_cff import microAODDefaultOutputCommand
-process.out = cms.OutputModule("PoolOutputModule", fileName = cms.untracked.string('myMicroAODOutputFile.root'),
+process.out = cms.OutputModule("PoolOutputModule", fileName = cms.untracked.string('myMicroAODOutput_test_JSON_read_SM.root'),
```
Comment (simonepigazzini) [inline MicroAOD/test/microAODstd.py]: Please rollback

File: MicroAOD/test/microAODstd.py [hunk 4]
```diff
@@ -11,21 +11,21 @@
...
 #    process.GlobalTag = GlobalTag(process.GlobalTag,'80X_dataRun2_2016LegacyRepro_v4','')
 #    process.source = cms.Source("PoolSource",fileNames=cms.untracked.vstring("/store/data/Run2016B/SingleElectron/MINIAOD/07Aug17_ver1-v1/110000/0248293E-578B-E711-A639-44A842CFC9D9.root"))
 
 ### 2017
-# process.GlobalTag = GlobalTag(process.GlobalTag,'','')
-# process.source = cms.Source("PoolSource",
-#                             fileNames=cms.untracked.vstring("/store/mc/RunIIFall17MiniAODv2/GluGluHToGG_M-125_13TeV_powheg_pythia8/MINIAODSIM/PU2017_12Apr2018_94X_mc2017_realistic_v14-v1/40000/0866D1A8-1941-E811-B61F-0CC47AF9B2E6.root"))
+#process.GlobalTag = GlobalTag(process.GlobalTag,'','')
+#process.source = cms.Source("PoolSource",
+#                             fileNames=cms.untracked.vstring("/store/mc/RunIIFall17MiniAODv2/GluGluToHHTo2B2G_node_SM_13TeV-madgraph/MINIAODSIM/PU2017_12Apr2018_94X_mc2017_realistic_v14-v2/20000/BA41D39E-8F7C-E811-A832-FA163E4A396F.root"))
```
Comment (simonepigazzini) [inline MicroAOD/test/microAODstd.py]: please rollback to GluGluH file

File: MicroAOD/test/microAODstd.py [hunk 5]
```diff
@@ -11,21 +11,21 @@
...
-process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32( 1000 ) )
+process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32( 10000 ) )
 process.MessageLogger.cerr.FwkReport.reportEvery = cms.untracked.int32( 100 )
 
 import os
 ### 2016
 process.GlobalTag = GlobalTag(process.GlobalTag, '', '')
 process.source = cms.Source("PoolSource",
-                             fileNames=cms.untracked.vstring("/store/mc/RunIISummer16MiniAODv3/VBFHToGG_M125_13TeV_amcatnlo_pythia8_v2/MINIAODSIM/PUMoriond17_94X_mcRun2_asymptotic_v3-v1/50000/38128C3C-892D-E911-AC8E-008CFA0087C4.root"))
+                             fileNames=cms.untracked.vstring("/store/mc/RunIISummer16MiniAODv3/GluGluToHHTo2B2G_node_SM_13TeV-madgraph/MINIAODSIM/PUMoriond17_94X_mcRun2_asymptotic_v3-v2/110000/CA97CD76-CD37-E911-ADBD-0090FAA57E64.root"))
```
Comment (simonepigazzini) [inline MicroAOD/test/microAODstd.py]: Please rollback to GluGluH file

File: MicroAOD/test/microAODstd.py (hunk 3 above)
Comment (panwarlsweet) [reply MicroAOD/test/microAODstd.py]: corrected.

File: MicroAOD/test/microAODstd.py (hunk 4 above)
Comment (panwarlsweet) [reply MicroAOD/test/microAODstd.py]: Corrected

File: MicroAOD/test/microAODstd.py (hunk 5 above)
Comment (panwarlsweet) [reply MicroAOD/test/microAODstd.py]: Corrected

File: README.md (hunk 2 above)
Comment (panwarlsweet) [reply README.md]: Done.

File: setup_flashgg.sh (hunk 1 above)
Comment (panwarlsweet) [reply setup_flashgg.sh:50]: Added proper explanation of DeepJet new training for this commit

**Comment by simonepigazzini** [comment]
Thank you. The PR looks good now. I will merge it before running the next set of test samples which will also include HH signal.

simone

**Comment by panwarlsweet** [comment]
Okay. Thanks a lot for all your help.

**Comment by panwarlsweet** [comment]
Hello Simone,

Sorry to bother you again but, could you please give me an estimated timeline when the samples with DeepJet should be made available? I need this info for the presentation on 13th May 2019 in Hgg meeting.

Thanks,

Lata Panwar

**Comment by simonepigazzini** [comment]
Dear Lata,

the production will probably start this week.

simone

**Comment by panwarlsweet** [comment]
Thanks, Simone. :)

**Pull request merged**